File: src/jugmeetup/scala/Sorts.scala
package jugmeetup.scala

import scala.annotation.tailrec

object Sorts extends App {
  @tailrec
  private def merge[A](lst1: List[A], lst2: List[A], merged: List[A])(lt: (A, A) => Boolean): List[A] =
    (lst1, lst2) match {
      case (_, Nil) => (lst1.reverse ::: merged).reverse
      case (Nil, _) => (lst2.reverse ::: merged).reverse
      case (h1 :: t1, h2 :: t2) => if (lt(h1, h2)) merge(t1, lst2, h1 :: merged)(lt) else merge(lst1, t2, h2 :: merged)(lt)
    }

  private def merge[A](lst1: List[A], lst2: List[A])(lt: (A, A) => Boolean): List[A] =
    (lst1, lst2) match {
      case (_, Nil) => lst1
      case (Nil, _) => lst2
      case (h1 :: t1, h2 :: t2) => if (lt(h1, h2)) h1 :: merge(t1, lst2)(lt) else h2 :: merge(lst1, t2)(lt)
    }

  def mergeSort[A](lst: List[A])(lt: (A, A) => Boolean): List[A] = lst match {
    case Nil => Nil
    case _ :: Nil => lst
    case _ =>
      val (left, right) = lst.splitAt(lst.length/2)
      merge(mergeSort(left)(lt), mergeSort(right)(lt), Nil)(lt)
  }
  
  println(mergeSort(List.fill(10)(math.random))(_ < _))
}

File: src/jugmeetup/java/Sorts.java
/*
 * Created on Sep 5, 2016
 */
package jugmeetup.java;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;

public class Sorts {
    private <A> List<A> merge(List<A> lst1,List<A> lst2,List<A> merged,Comparator<A> comp) {
        if(lst1.isEmpty()) {
            merged.addAll(lst2);
            return merged;
        } else if(lst2.isEmpty()) {
            merged.addAll(lst1);
            return merged;
        } else if(comp.compare(lst1.get(0), lst2.get(0)) < 0) {
            merged.add(lst1.get(0));
            return merge(lst1.subList(1, lst1.size()), lst2, merged, comp);
        } else {
            merged.add(lst2.get(0));
            return merge(lst1, lst2.subList(1, lst2.size()), merged, comp);
        }
    }

    public <A> List<A> mergeSort(List<A> lst, Comparator<A> comp) {
        if(lst.size() < 2) return lst;
        else {
            List<A> left = lst.subList(0, lst.size()/2-1);
            List<A> right = lst.subList(lst.size()/2, lst.size());
            return merge(mergeSort(left,comp),mergeSort(right,comp), new ArrayList<A>(),comp);
        }
    }
}
